### DISCUSSION TITLE
LFE smart cells for Elixir Livebooks

### DISCUSSION THREAD

**Discussion by gliderproarc** [lfe/lfe#464] [General]
I've recently been learning how to code and of the languages I have tried so far, LFE holds a special place in my heart. I would love to see more people learn about LFE and how fun Lisp is, and how well it pairs with the BEAM.

I've also been learning Elixir, and while I do most of my code editing in Emacs, having something like a computational notebook has been really fun. I already use LFE in a similar way in Emacs Org mode, mixing prose and LFE code, but I realize a lot of people don't want to bother with Emacs. Elixir's LiveBooks are a very approachable way of letting people just getting started to have a nice time trying out stuff like LFE.

I have been toying around with the idea of if it would be possible to make something like a smart cell for LFE code in an Elixir Livebook. From what I have seen so far, these smart cells are just Elixir code, but something tells me it won't be so easy to just in-line LFE in the middle of a Elixir document.

But if it were possible to do something like call a Mix task or something to compile a string of LFE code, pass it the context of where it is in execution of the document, run it with the current environment, then pass it back to the LiveBook, I can imagine how nice it would be to use the format of a LiveBook, but write LFE into these smart cells. The basis of how it works could probably be used for other beam langue's like Gleam to one day also have their own smart cell. 

My questions is, am I missing some kind of fundamental problem to trying to make this happen? 

I took a look at the GitHub discussions section for live book project, and it looks like it should be possible since there was talk of making a smart cell for running Erlang code. But before I dive in too deep, I am curious if any one thinks this could end up being trickier than it seems. I have been looking at how LFE gets executed in an Org-mode code cell in Emacs and imagining it would be kind of similar, just a bit trickier when it comes to the environment (since LiveBooks don't pass variables in by name like Org-mode code cells. And it might not be possible to just write code out to a temporary text file to be trashed later).

**Comment by oubiwann** [comment]
This shouldn't be that tricky for simple things, namely, any code that exists in Erlang/OTP proper or LFE. Depending upon how Livebooks work, you could either compile or eval. If Livebooks maintain some sort of state so that a definition in one cell could be used in another, then thinks might get a little tricky. LFE handles this with a classic Lisp "environment" ...

We've done various integrations in the past with LFE eval'ing and compiling, so there's a good chance that everything you need is ready to go (again, it depends upon the exact requirements of Livebooks on how third-party code is expected to be handled). If LFE is missing some capability, there's a good chance we could add a new feature/function to support the use case :-)

**Reply by gliderproarc** [reply]
Thank you for the reply!

That's a really reassuring to hear! If I could just get the basics working, I think someone with more experience could add to it later and add functionality for use cases where it's not the simple stuff. 

I also heard that Erlang was just added as a language to LiveBooks, so I bet that work will serve as a great starting point for seeing what needs to be done and where to "plug-in" to LiveBooks's inner workings.
